src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,548 @@
+unsafe impl bytemuck::Pod for CBufferContents {}
+
+// fn buffer_set_contents(buffer: u32, cbuffer: &CBufferContents) -> bool;
+// Copy the buffer_contents structure back into the host shaping engine's buffer. This should typically be called at the end of shaping.
+fn buffer_set_contents(mut caller: Caller<'_, ShapingData>, _buffer: u32, cbuffer: u32) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut buffer = [0; core::mem::size_of::<CBufferContents>()];
+    let Ok(()) = memory.read(caller.as_context_mut(), cbuffer as usize, &mut buffer) else {
+        return 0;
+    };
+    let Ok(buffer) = bytemuck::try_from_bytes::<CBufferContents>(&buffer) else {
+        return 0;
+    };
+
+    let (mem_data, store_data) = memory.data_and_store_mut(&mut caller);
+
+    let array_length = buffer.length as usize * core::mem::size_of::<hb_glyph_info_t>();
+
+    store_data.buffer.len = buffer.length as usize;
+
+    store_data.buffer.info.clear();
+    store_data
+        .buffer
+        .info
+        .extend_from_slice(bytemuck::cast_slice(
+            &mem_data[buffer.info as usize..buffer.info as usize + array_length],
+        ));
+
+    store_data.buffer.pos.clear();
+    store_data
+        .buffer
+        .pos
+        .extend_from_slice(bytemuck::cast_slice(
+            &mem_data[buffer.position as usize..buffer.position as usize + array_length],
+        ));
+
+    1
+}
+
+#[allow(unused)]
+// fn debugprint(s: *const u8);
+// Produces a debugging message in the host shaper's log output;
+fn debugprint(mut caller: Caller<'_, ShapingData>, s: u32) {
+    // rustybuzz does not have a log output.
+    // internal_debug(caller, s) // <--- uncomment this if you need to print the module messages.
+}
+
+#[allow(unused)]
+fn internal_debug(mut caller: Caller<'_, ShapingData>, s: u32) {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    if memory.data(&caller).get(s as usize).is_none() {
+        return;
+    }
+    let bytes = &memory.data(&caller)[s as usize..];
+    let msg = CStr::from_bytes_until_nul(bytes)
+        .unwrap_or_default()
+        .to_string_lossy();
+
+    std::eprintln!("{msg}");
+}
+
+// fn shape_with(font: u32, buffer: u32, features: u32, num_features: u32, shaper: *const u8) -> i32;
+// Run another shaping engine's shaping process on the given font and buffer. The only shaping engine guaranteed to be available is ot, the OpenType shaper, but others may also be available. This allows the WASM author to process a buffer "normally", before further manipulating it.
+fn shape_with(
+    mut caller: Caller<'_, ShapingData>,
+    // We don't use font token and buffer token
+    _font: u32,
+    _buffer: u32,
+    // harfbuzz-wasm doesn't use the features pointer
+    _features: u32,
+    _num_features: u32,
+    // we dont have custom shapers (yet?).
+    _shaper: u32,
+) -> i32 {
+    // potentially we could read the shaper pointed to by `shaper`
```

Comment:
Let's add the `log` dependency and warn when non `ot` shaper is requested.
Not sure how to report an error from this method. I guess simply log and skip the shape call.